modules/videoio/test/test_gstreamer.cpp

```diff
@@ -0,0 +1,320 @@
+//
+//         if (flag_rect)
+//         {
+//             Point2d inverse_first, inverse_last;
+//             for (int x = first.x; x < width; x++)
+//             {
+//                 if (frame.at<uchar>(first.y, x) == 255)
+//                 {
+//                     inverse_first.x = x;
+//                 }
+//             }
+//             inverse_first.y = first.y;
+//
+//             for (int y = first.y; y < height; y++)
+//             {
+//                 if (frame.at<uchar>(y, first.x) == 255)
+//                 {
+//                     inverse_last.y = y;
+//                 }
+//             }
+//             inverse_last.x = first.x;
+//
+//             points = std::vector<Point2d>();
+//             points.push_back(inverse_first);
+//             points.push_back(inverse_last);
+//         }
+//         return points;
+//     }
+//
+// protected:
+//     string format, pattern;
+//     Size size;
+//     Size size_nv12;
+//     int count_frames;
+//     std::ostringstream pipeline;
+// };
+//
+// string color_spaces[]  = {
+//     string("BGR"),  string("GRAY8"), string("UYVY"),
+//     string("YUY2"), string("YVYU"),  string("NV12"),
+//     string("NV21"), string("YV12"),  string("I420"),
+//     string("video/x-bayer"), string("image/jpeg")
+// };
+//
+// string pattern_video[]  = {
+//     string("ball"), string("circular"),
+//     string("bar"),  string("gradient")
+// };
+//
+// TEST_P(Videoio_Gstreamer_Test, test_size_formats) { testSizeFormats(); }
+// TEST_P(Videoio_Gstreamer_Test, test_object_structure) { testObjectStructure(); }
+//
+// INSTANTIATE_TEST_CASE_P(videoio, Videoio_Gstreamer_Test,
+//                         testing::Combine(
+//                             testing::ValuesIn(color_spaces),
+//                             testing::ValuesIn(pattern_video)));
+
+typedef cv::Point_< double > Point2d;
+typedef tuple< string, Size, Size, int > Param;
+typedef testing::TestWithParam< Param > Videoio_Gstreamer_Test;
+
+TEST_P(Videoio_Gstreamer_Test, test_object_structure)
+{
+    string format    = get<0>(GetParam());
+    Size frame_size  = get<1>(GetParam());
+    Size mat_size    = get<2>(GetParam());
+    int convertToRGB = get<3>(GetParam());
+    int count_frames = 30;
+    std::ostringstream pipeline;
+    pipeline << "videotestsrc pattern=ball num-buffers=" << count_frames << " ! " << format;
+    pipeline << ", width=" << frame_size.width << ", height=" << frame_size.height << " ! appsink";
+    VideoCapture cap(pipeline.str(), CAP_GSTREAMER);
+
+    ASSERT_TRUE(cap.isOpened());
+    for (int i = 0; i < count_frames; ++i)
+    {
+        Mat buffer, frame, rgb_frame, binary_frame;
+        cap >> buffer;
+        frame = (format == "jpegenc ! image/jpeg") ? imdecode(buffer, IMREAD_UNCHANGED) : buffer;
+        EXPECT_EQ(mat_size, frame.size());
+
+        cvtColor(frame, rgb_frame, convertToRGB);
+        cvtest::threshold(rgb_frame, binary_frame, 128, 255, CV_THRESH_BINARY);
```

Comment:
`cvtest::threshold` expects preallocated image as destination. Initialize it like this: `Mat binary_frame(rgb_frame.size(), rgb_frame.type());`